### PR TITLE
mgr: register OSDs in ms_handle_accept

### PR DESCRIPTION
**Pull request by batrick**
It's a no-no to acquire locks in these "fast" messenger methods. This can lead to messenger slow downs in the best case as it's blocking reads on the wire. In the worse case, the messenger may deadlock with other threads, preventing any further message reads off the wire.

It's not obvious this method is "fast" so I've added a comment regarding this.

Fixes: https://tracker.ceph.com/issues/61874

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
Hmm, we should audit other users of ms_handle_accept -- I haven't looked at this in a while but im pretty sure we put "fast" in the name of everything which needed to be non-blocking when we set that interface up, so there may be a conflict here.

**Comment by batrick** [comment]
> Hmm, we should audit other users of ms_handle_accept -- I haven't looked at this in a while but im pretty sure we put "fast" in the name of everything which needed to be non-blocking when we set that interface up, so there may be a conflict here.

The issue is `ms_handle_authentication` not `ms_handle_accept`. The former is run as a "fast" method and the latter in the regular dispatcher threads.

I've already gone through other users of `ms_handle_authentication` but have now added a commit that makes it more clear in the method name, as with the others. PTAL.

**Comment by batrick** [comment]
ping @rzarzynski

**Review by vshankar — COMMENTED** [review]
LGTM @batrick - mgr/DaemonServer.cc source changes are really meaningful!

Needs @rzarzynski approval though.

**Review by rzarzynski — APPROVED** [review]
This looks sane.

I very like how the refactor and fix have been split apart. It should make backporting straightforward. Thanks!

File: src/msg/Dispatcher.h [hunk 1]
```diff
@@ -204,13 +204,16 @@ class Dispatcher {
...
    *
-   * Authenticated result/state will be attached to the Connection.
+   * Authenticated result/state will be attached to the Connection. This is
+   * called via the MonClient.
+   *
+   * Do not acquire locks in this method! It is considered "fast" delivery.
    *
    * return 1 for success
    * return 0 for no action (let another Dispatcher handle it)
    * return <0 for failure (failure to parse caps, for instance)
```
Comment (idryomov) [inline src/msg/Dispatcher.h:214]: > I've already gone through other users of `ms_handle_authentication` [...]

Have you looked into why most callers seem to ignore the failure?

Comment (batrick) [reply src/msg/Dispatcher.h:214]: That's a great question. I did not. It appears the only side-effect is that caps which fail to parse can continue talking to the daemon but e.g. for the mons it just treats it as unauthenticated with no privileges.

I will create a new ticket to clean this up. I don't think it necessarily needs to be in this PR.

Comment (batrick) [reply src/msg/Dispatcher.h:214]: https://tracker.ceph.com/issues/62382

Comment (batrick) [reply src/msg/Dispatcher.h:214]: I will work on that tomorrow.

File: src/osd/OSD.h [hunk 2]
```diff
@@ -1526,7 +1526,7 @@ class OSD : public Dispatcher,
     bool ms_handle_refused(Connection *con) override {
       return osd->ms_handle_refused(con);
     }
-    int ms_handle_authentication(Connection *con) override {
+    int ms_handle_fast_authentication(Connection *con) override {
       return true;
```
[the remark this answers is not in the document]
Comment (batrick) [reply src/osd/OSD.h:1530]: I'll fix this too in https://tracker.ceph.com/issues/62382

**Comment by ljflores** [comment]
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomci87i4GUf1826-wip-yuri10-testing-2023-08-17-1444-old-wip-yuri10-testing-2023-08-15-1601-old-wip-yuri10-testing-2023-08-15-1009

**Comment by batrick** [comment]
> Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrellocomci87i4GUf1826-wip-yuri10-testing-2023-08-17-1444-old-wip-yuri10-testing-2023-08-15-1601-old-wip-yuri10-testing-2023-08-15-1009

Thanks Laura!

**Pull request merged**